studio.py

```diff
@@ -14,14 +22,211 @@
     "Fenix": Fenix,
     "Fennec": Fennec,
     "Refbrow": RefBrow,
-    "Chrome": Chrome
+    "Chrome": Chrome,
 }
-PROXYS = {"mitm2": MITMProxy202, "mitm": MITMProxy404, "wpr": WebPageReplay}
+
+PROXYS = {"mitm2": MITMProxy202,
+          "mitm4": MITMProxy404,
+          "wpr": WebPageReplay}
+
+RECORD_TIMEOUT = 30
+
+
+class Mode:
+    def __init__(self, app, app_path, proxy, certutil, path, sites, url):
+        self.app = app
+        self.app_path = app_path
+        self.proxy = proxy
+        self.certutil = certutil
+        self.path = path
+        self.sites = sites
+        self.url = url
+        self.information = {}
+
+    def _digest_file(self, file, algorithm):
+        """I take a file like object 'f' and return a hex-string containing
+        of the result of the algorithm 'a' applied to 'f'."""
+        with open(file, "rb") as f:
+            h = hashlib.new(algorithm)
+            chunk_size = 1024 * 10
+            data = f.read(chunk_size)
+            while data:
+                h.update(data)
+                data = f.read(chunk_size)
+            name = repr(f.name) if hasattr(f, "name") else "a file"
+            print("hashed %s with %s to be %s" % (name, algorithm, h.hexdigest()))
+            return h.hexdigest()
+
+
+    def replaying(self):
+        with PROXYS[self.proxy](path=self.path, mode="replay") as proxy_service:
+            app_service = APPS[self.app](proxy_service, self.certutil)
+            app_service.start(self.url)
+
+    def recording(self):
+        print("Starting record mode!!!")
+        if not os.path.exists(self.path):
+            print("Creating recording path: %s" % self.path)
+            os.mkdir(self.path)
+
+        recordings = self.parse_sites_json()
+
+        for recording in recordings:
+            if not os.path.exists(os.path.dirname(recording["recording_path"])):
+                print(
+                        "Creating recording path: %s"
+                        % os.path.dirname(recording["recording_path"])
+                )
+                os.mkdir(os.path.dirname(recording["recording_path"]))
+
+            if not recording.get("login", False):
+                with PROXYS[self.proxy](
+                        path=recording["recording_path"], mode="record"
+                ) as proxy_service:
+                    app_service = APPS[self.app](
+                        proxy_service, self.certutil, self.app_path
+                    )
+                    app_service.start(recording["url"])
+
+                    if not recording.get("login", None):
```

Comment:
We're already inside a block for non-logged in sites.

Reply:
Yep removed one block